From 79f08ff39022f61677cc5963b412581025bf5c1b Mon Sep 17 00:00:00 2001
From: Xufeng Zhang <xufeng.zhang@windriver.com>
Date: Tue, 28 Jun 2011 16:40:11 +0800
Subject: [PATCH 10/47] talitos: move kmem_cache_create before registering crypto algorithms to avoid kernel panic

The panic messages are as below

Faulting instruction address: 0xc00b23c4
Oops: Kernel access of bad area, sig: 11 [#1] SMP NR_CPUS=2 P2020 DS last sysfs file:
Modules linked in:
NIP: c00b23c4 LR: c035c5c0 CTR: c035ae3c
REGS: ef8f1c10 TRAP: 0300   Not tainted  (2.6.35-rc5-00105-g089862e)
MSR: 00021000 <ME,CE>  CR: 22044022  XER: 00000000
DEAR: 00000000, ESR: 00000000
TASK = ef8f72e0[1208] 'cryptomgr_test' THREAD: ef8f0000 CPU: 1
GPR00: 00000004 ef8f1cc0 ef8f72e0 00000000 000000f1 ef8a1c00 00000001 00000000
GPR08: 00000000 c0602530 00000400 01665000 00000000 38803320 7ff9a900 00000000
GPR16: 00000000 7ff8ef40 ef8f1e1c ec94b000 ef8f1ddc ef8f34a0 ef8f1da8 00000000
GPR24: ec91f300 00000000 00000000 00000000 00000000 000000f1 00029000 00000000
NIP [c00b23c4] kmem_cache_alloc+0x40/0xb8
LR [c035c5c0] crypto_edesc_alloc+0x60/0x64
Call Trace:
[ef8f1ce0] [c035c5c0] crypto_edesc_alloc+0x60/0x64
[ef8f1cf0] [c035c730] talitos_edesc_alloc+0x16c/0x294
[ef8f1d20] [c035efd4] ahash_process_req+0x224/0x598
[ef8f1d80] [c01d988c] crypto_ahash_op+0x38/0xf8
[ef8f1da0] [c01dc310] test_hash+0x158/0x52c
[ef8f1f10] [c01dc72c] alg_test_hash+0x48/0xac
[ef8f1f30] [c01de048] alg_test+0x7c/0x1f4
[ef8f1fa0] [c01db554] cryptomgr_test+0x4c/0x54
[ef8f1fb0] [c005ac48] kthread+0x78/0x7c
[ef8f1ff0] [c0010244] kernel_thread+0x4c/0x68
---[ end trace d703b0766209fbd0 ]---

Extracted from vendor drop QorIQ-NONDPAA-SDK-V1-20110429_ltib.iso
linux-2.6.35-qoriq-talitos-move-kmem_cache_create-before-r.patch.

Signed-off-by: Jin Qing <b24347@freescale.com>
Integrated-by: Xufeng Zhang <xufeng.zhang@windriver.com>
---
 drivers/crypto/talitos.c |   13 ++++++++++---
 1 files changed, 10 insertions(+), 3 deletions(-)

diff --git a/drivers/crypto/talitos.c b/drivers/crypto/talitos.c
index 20ebefd..3a9df4f 100644
--- a/drivers/crypto/talitos.c
+++ b/drivers/crypto/talitos.c
@@ -2587,6 +2587,16 @@ static int talitos_probe(struct of_device *ofdev,
 			dev_info(dev, "hwrng\n");
 	}
 
+	priv->netcrypto_cache = kmem_cache_create("netcrypto_cache",
+						MAX_DESC_LEN, 0,
+					SLAB_HWCACHE_ALIGN, NULL);
+	if (!priv->netcrypto_cache) {
+		printk(KERN_ERR "%s: failed to create block cache\n",
+				__func__);
+		err = -ENOMEM;
+		goto err_out;
+	}
+
 	/*
 	 * register with async_tx xor, if capable
 	 * SEC 2.x support up to 3 RAID sources,
@@ -2629,9 +2639,6 @@ static int talitos_probe(struct of_device *ofdev,
 			}
 		}
 	}
-	priv->netcrypto_cache = kmem_cache_create("netcrypto_cache",
-						MAX_DESC_LEN, 0,
-					SLAB_HWCACHE_ALIGN, NULL);
 
 	return 0;
 
-- 
1.7.0.4

